### PR TITLE
feat: add `user-agent` to templates

### PR DESCRIPTION
**Pull request by everly-gif**
add `user-agent` to templates

### CODE REVIEW

**Comment by lohanidamodar** [comment]
@everly-gif are you working on these?

**Comment by everly-gif** [comment]
> @everly-gif are you working on these?

Yes, synced and tests are fixed, pls review and merge

File: templates/dart/lib/src/client_io.dart.twig [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import 'dart:io';
+import 'dart:io' show Platform;
```
Comment (lohanidamodar) [inline templates/dart/lib/src/client_io.dart.twig:2]: we don't need to import it again
```suggestion
```

File: templates/dart/pubspec.yaml.twig [hunk 2]
```diff
@@ -10,6 +10,7 @@ environment:
 dependencies:
   http: ^0.13.5
 
+
```
Comment (lohanidamodar) [inline templates/dart/pubspec.yaml.twig:13]:
```suggestion
```

File: templates/node/lib/client.js.twig [hunk 3]
```diff
@@ -11,6 +12,7 @@ class Client {
         this.headers = {
             'accept-encoding': '*',
             'content-type': '',
+            'user-agent' : `{{spec.title | caseUcfirst}}Node/{{ sdk.version }} (${os.type()} ${os.version()}; ${os.arch()})`,
```
Comment (lohanidamodar) [inline templates/node/lib/client.js.twig:15]:
```suggestion
            'user-agent' : `{{spec.title | caseUcfirst}}{{language.name | caseUcFirst}}SDK/{{ sdk.version }} (${os.type()} ${os.version()}; ${os.arch()})`,
```

File: src/SDK/SDK.php [hunk 4]
```diff
@@ -103,6 +103,9 @@ public function __construct(Language $language, Spec $spec)
         $this->twig->addFilter(new TwigFilter('caseUcfirst', function ($value) {
             return ucfirst($this->helperCamelCase($value));
         }));
+        $this->twig->addFilter(new TwigFilter('deviceInfo', function ($value) {
```
Comment (lohanidamodar) [inline src/SDK/SDK.php:106]: can we move this filter to PHP instead, as it's only used in PHP language.

**Review by lohanidamodar — CHANGES_REQUESTED** [review]

File: templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig [hunk 5]
```diff
@@ -62,11 +62,13 @@ class Client @JvmOverloads constructor(
     init {
         headers = mutableMapOf(
             "content-type" to "application/json",
+            "user-agent" to "{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/{{ sdk.version }} ${System.getProperty("http.agent")}",
             "x-sdk-name" to "{{ sdk.name }}",
             "x-sdk-platform" to "{{ sdk.platform }}",
             "x-sdk-language" to "{{ language.name | caseLower }}",
             "x-sdk-version" to "{{ sdk.version }}"{% if spec.global.defaultHeaders | length > 0 %},{% endif %}
             
+
```
Comment (stnguyen90) [inline templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig:71]: Is this an extra unnecessary newline?

File: templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig [hunk 6]
```diff
@@ -62,11 +62,13 @@ class Client @JvmOverloads constructor(
     init {
         headers = mutableMapOf(
             "content-type" to "application/json",
+            "user-agent" to "{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/{{ sdk.version }} ${System.getProperty("http.agent")}",
```
Comment (stnguyen90) [inline templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig:65]: Should `sdk.version` be in this?

File: templates/node/lib/client.js.twig [hunk 7]
```diff
@@ -11,6 +12,7 @@ class Client {
         this.headers = {
             'accept-encoding': '*',
             'content-type': '',
+            'user-agent' : `{{spec.title | caseUcfirst}}{{language.name | caseUcFirst}}SDK/{{ sdk.version }} (${os.type()} ${os.version()}; ${os.arch()})`,
```
Comment (stnguyen90) [inline templates/node/lib/client.js.twig:15]: Should `sdk.version` be in this?

File: templates/python/package/client.py.twig [hunk 8]
```diff
@@ -11,6 +11,7 @@ class Client:
         self._endpoint = '{{spec.endpoint}}'
         self._global_headers = {
             'content-type': '',
+            'user-agent' : '{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/{{ sdk.version }} (${os.uname().sysname}; ${os.uname().version}; ${os.uname().machine})',
```
Comment (stnguyen90) [inline templates/python/package/client.py.twig:14]: should `sdk.version` be in this?

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

File: templates/deno/src/client.ts.twig [hunk 9]
```diff
@@ -11,6 +11,7 @@ export class Client {
     endpoint: string = '{{spec.endpoint}}';
     headers: Payload = {
         'content-type': '',
+        'user-agent' : `{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/{{ sdk.version }} (${Deno.build.os}; ${Deno.build.arch})`,
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply templates/deno/src/client.ts.twig:14]: @stnguyen90 not sure what you mean?

File: templates/dart/lib/src/client_io.dart.twig [hunk 10]
```diff
@@ -43,8 +43,10 @@ class ClientIO extends ClientBase with ClientMixin {
       'x-sdk-platform': '{{ sdk.platform }}',
       'x-sdk-language': '{{ language.name | caseLower }}',
       'x-sdk-version': '{{ sdk.version }}',
+      'user-agent' : '{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/({{Platform.operatingSystem}}; {{Platform.operatingSystemVersion}})',
 {% for key,header in spec.global.defaultHeaders %}
       '{{key}}' : '{{header}}',
+
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply templates/dart/lib/src/client_io.dart.twig:49]: fixed

File: templates/dotnet/src/Appwrite/Client.cs.twig [hunk 11]
```diff
@@ -30,6 +30,7 @@ namespace {{ spec.title | caseUcfirst }}
             this.headers = new Dictionary<string, string>()
             {
                 { "content-type", "application/json" },
+                { "user-agent" , "{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/{{ sdk.version }} (${Environment.OSVersion.Platform}; ${Environment.OSVersion.VersionString})"}{% if spec.global.defaultHeaders | length > 0 %},{% endif %}
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply templates/dotnet/src/Appwrite/Client.cs.twig:33]: I think sdk version is fine, fixed the comma issue

File: templates/go/client.go.twig [hunk 12]
```diff
@@ -65,6 +65,7 @@ func NewClient() Client {
 	headers := map[string]string{
 {% for key,header in spec.global.defaultHeaders %}
 		"{{key}}" : "{{header}}",
+		"user-agent" : "{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/ ({{deviceInfo}})",
```
[the remark this answers is not in the document]
Comment (lohanidamodar) [reply templates/go/client.go.twig:68]: seems such, removed and added `sdk.version` instead. I don't have enough go knowledge for now to add other device info

File: templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig (hunk 5 above)
Comment (lohanidamodar) [reply templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig:71]: removed

File: templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig (hunk 6 above)
Comment (lohanidamodar) [reply templates/kotlin/src/main/kotlin/io/appwrite/Client.kt.twig:65]: I think it was discussed and agreed to keep

**Comment by lohanidamodar** [comment]
@stnguyen90 I think the SDK version makes sense in the user-agent.

File: templates/dart/lib/src/client_io.dart.twig [hunk 13]
```diff
@@ -43,6 +43,7 @@ class ClientIO extends ClientBase with ClientMixin {
       'x-sdk-platform': '{{ sdk.platform }}',
       'x-sdk-language': '{{ language.name | caseLower }}',
       'x-sdk-version': '{{ sdk.version }}',
+      'user-agent' : '{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/({{Platform.operatingSystem}}; {{Platform.operatingSystemVersion}})',
```
Comment (stnguyen90) [inline templates/dart/lib/src/client_io.dart.twig:46]: Missing sdk version.

File: templates/go/client.go.twig [hunk 14]
```diff
@@ -65,6 +65,7 @@ func NewClient() Client {
 	headers := map[string]string{
 {% for key,header in spec.global.defaultHeaders %}
 		"{{key}}" : "{{header}}",
+		"user-agent" : "{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/{{ sdk.version }}",
```
Comment (stnguyen90) [inline templates/go/client.go.twig:68]: In go, you can use `runtime.GOOS`:

```go
package main

import (
	"fmt"
	"runtime"
)

func main() {
	s := fmt.Sprintf("(%s;%s)", runtime.GOOS, runtime.GOARCH)
	fmt.Println(s)
}
```

outputs: 

```
(linux;amd64)
```

https://go.dev/play/p/aS0Ik4UnKXj

File: templates/node/lib/client.js.twig (hunk 7 above)
Comment (stnguyen90) [inline templates/node/lib/client.js.twig:15]: should the ; be after the os type rather than after the version?

File: src/SDK/Language/PHP.php [hunk 15]
```diff
@@ -384,7 +384,10 @@ public function getFilters(): array
         return [
             new TwigFilter('getReturn', function ($value) {
                 return $this->getReturn($value);
-            })
+            }),
+            new TwigFilter('deviceInfo', function ($value) {
+                return php_uname('s') . ';' . php_uname('v') . ';' . php_uname('m');
```
Comment (stnguyen90) [inline src/SDK/Language/PHP.php:389]: For PHP, every value is only separated by `;` and no spaces. This isn't consistent between the various user agents.

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

File: templates/dart/lib/src/client_io.dart.twig (hunk 13 above)
Comment (lohanidamodar) [reply templates/dart/lib/src/client_io.dart.twig]: updated

File: templates/node/lib/client.js.twig (hunk 7 above)
Comment (lohanidamodar) [reply templates/node/lib/client.js.twig]: updated

File: src/SDK/Language/PHP.php (hunk 15 above)
Comment (lohanidamodar) [reply src/SDK/Language/PHP.php:389]: this part is just os name, type and arch separated by `;` same as other SDKs?

File: templates/go/client.go.twig (hunk 14 above)
Comment (lohanidamodar) [reply templates/go/client.go.twig]: updated

**Comment by abnegate** [comment]
@lohanidamodar Do we want to update the client SDKs `user-agent` header format? They will be inconsistent with these otherwise

**Comment by lohanidamodar** [comment]
> @lohanidamodar Do we want to update the client SDKs `user-agent` header format? They will be inconsistent with these otherwise

Only difference will be, in the client we want to use the package info of the app instead of SDK/version right. That's intended. For server side SDKs apps package, version is no much of importance

File: src/SDK/Language/PHP.php (hunk 15 above)
Comment (stnguyen90) [reply src/SDK/Language/PHP.php:389]: The other SDKs have it separated by `; ` (extra space) rather than just `;`

**Review by stnguyen90 — CHANGES_REQUESTED** [review]

**Comment by abnegate** [comment]
> > @lohanidamodar Do we want to update the client SDKs `user-agent` header format? They will be inconsistent with these otherwise
> 
> Only difference will be, in the client we want to use the package info of the app instead of SDK/version right. That's intended. For server side SDKs apps package, version is no much of importance

True makes sense 👌

File: templates/node/lib/client.js.twig [hunk 16]
```diff
@@ -11,6 +12,7 @@ class Client {
         this.headers = {
             'accept-encoding': '*',
             'content-type': '',
+            'user-agent' : `{{spec.title | caseUcfirst}}{{language.name | caseUcFirst}}SDK/{{ sdk.version }} (${os.type()}; ${os.version()}; ${os.arch()})`,
```
Comment (abnegate) [inline templates/node/lib/client.js.twig:15]:
```suggestion
'user-agent' : `{{spec.title | caseUcfirst}}{{language.name | caseUcfirst}}SDK/{{ sdk.version }} (${os.type()}; ${os.version()}; ${os.arch()})`,
```

File: templates/go/client.go.twig [hunk 17]
```diff
@@ -65,6 +65,7 @@ func NewClient() Client {
 	headers := map[string]string{
 {% for key,header in spec.global.defaultHeaders %}
 		"{{key}}" : "{{header}}",
+		"user-agent" : fmt.Sprintf("{{spec.title | caseUcfirst}}{{ language.name | caseUcfirst }}SDK/{{ sdk.version }} (%s; %s)", runtime.GOOS, runtime.GOARCH),
```
Comment (abnegate) [inline templates/go/client.go.twig:68]: We'll need to add `"runtime"` to the imports at the top

Comment (lohanidamodar) [reply templates/go/client.go.twig:69]: updated

**Review by abnegate — APPROVED** [review]

**Review by stnguyen90 — APPROVED** [review]

**Pull request merged**